src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+        eventBufferOptions: defaultEventBufferOptions,
+      },
+    );
+
+    context.model = model;
+    context.channelName = channelName;
+    context.ably = ably;
+
+    context.eventData = {
+      name: 'update',
+      data: {
+        noOfMessages: 34,
+      },
+      mutationId: 'some-id-1',
+    };
+  });
+
+  afterEach<TestContext>(async ({ model }) => {
+    await model.dispose();
+    vi.restoreAllMocks();
+  });
+
+  it<TestContext>('changes state on sync, pause, resume, dispose', async ({ model }) => {
+    expect(model.state).toEqual('initialized');
+
+    const sync = model.sync({
+      data: {
+        foo: 'foobar',
+        bar: {
+          baz: 1,
+        },
+      },
+      sequenceId: '1',
+    });
+    expect(model.state).toEqual('syncing');
+
+    await sync;
+    expect(model.state).toEqual('ready');
+
+    await model.pause();
+    expect(model.state).toEqual('paused');
+
+    await model.resume();
+    expect(model.state).toEqual('ready');
+
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
+
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
+    await confirmation;
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(3);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(3, null, eventData.data);
+    expect(model.data.confirmed).toEqual(eventData.data);
+  });
+
+  it<TestContext>('rejects the data in optimistic() and rolls back the changes if back-end rejects the data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    model.subscribe(subscriptionSpy);
+
+    await model.optimistic(eventData);
+    expect(model.data.optimistic).toEqual(eventData.data);
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-reject': true,
+        },
+      },
+    });
+
+    await subscriptionCalls[1];
```

Comment:
We should still await `subscriptionCalls[0]` and assert initial state, and assert that the `confirmation` promise returned by the `optimistic` call rejects.

Reply:
have a look, did I do it as you meant it?